Tree of Thought introducing a new ToTChain.

vadimgu opened the pull request:
# [WIP] Tree of Thought introducing a new ToTChain.

This PR adds a new chain called ToTChain that implements the ["Large Language Model Guided Tree-of-Though"](https://arxiv.org/pdf/2305.08291.pdf) paper.

There's a notebook example `docs/modules/chains/examples/tot.ipynb` that shows how to use it.


Implements #4975


## Who can review?

Community members can review the PR once tests pass. Tag maintainers/contributors who might be interested:

- @hwchase17
- @vowelparrot

vgubergrits committed:
Implement the Tree of Thout (ToT) Chain

vgubergrits committed:
Adding notebook example for ToT

vgubergrits committed:
Fixing the memory stack for ToT

vgubergrits committed:
ToT test using a sudoku.

vgubergrits committed:
ToT Linting and other style adjustments.

hwchase17 commented on langchain/chains/tot/base.py:
```diff
@@ -0,0 +1,317 @@
+        run_manager: Optional[CallbackManagerForChainRun] = None,
+    ) -> Optional[str]:
+        """Algorithm 2 from the ToT paper."""
+        checker = partial(self.checker, problem_description)
```
nit: not sure these lines add much

vadimgu replied:
I was trying to write something that looked very close to the algorithm in the paper. But I agree that it doesn't make it more readable.

hwchase17 commented on langchain/chains/tot/base.py:
```diff
@@ -0,0 +1,317 @@
+        prompt = prompter(problem_description)
+        for _ in range(self.k):
+            self.log_request(prompt, run_manager)
+            response = self.parse_llm_output(llm(prompt))
```
we should construct and use LLMChain here - will help with logging (may make log_request and log_response unneccessary

hwchase17 reviewed (commented):
awesome! left some stylistic comments, the one about using LLMChain is the main one

is there a better example? right now in the notebook seems to only use one iteration?

vgubergrits committed:
Addressing comments in the PR

- Using LLMChain instead of a bare language model
- Moving the parser into langchain/chains/tot/prompts.py
- Adjusting the notebook demonstrating ToT.

vgubergrits committed:
Make ToT chain compatible with Python3.8

vadimgu commented:
I test it with Python3.8 and removed a listing issue with an unused import.

vadimgu commented:
Do you think that ToTChecker should be a chain?

vgubergrits committed:
ToT Add thought strategy

Adding a thought strategy for thought generation.

1. SampleCoTStrategy
2. ProposePromptStrategy

levalencia commented:
is this already in the latest available version?

hwchase17 committed:
cr

hwchase17 committed:
stash

hwchase17 committed:
cr

hwchase17 committed:
cr
